File: 发给海前/main/文件/EXR2ISET/piReadEXR.m
%这就是进行解包，全是依赖EXR2MAT的
function output = piReadEXR(filename, varargin)
%% Read multispectral data from a .exr file (Openexr format)
%
%   [imageData, imageSize, lens] = piReadEXR(filename)
%
%% Deprecated in favor of python wrapper version
%% Left in tree for anyone who has OpenEXR via .mex files working
%% But has a problem with python + pyexr -- David Cardinal 10/05/21
%% Using docker to convert exr to mat file. -- Zhenyi 10/14/21
%
% Required Input
%   filename - existing .exr file
%
% Optional parameter/val
%   dataType - specify the type of data for the output
%   NSpectrumSamples - number of spectrum samples
%
% Returns
%  
% Read .exr image data from the fiven filename.
%
%
% Zhenyi, 2020
% 
% Make sure openexr functions are compiled: 
%               (run make.m in ~/iset3d/ext/openexr)
% 
% Get exr channel information
%{
      data = exrinfo(filename);
      fprintf( '%s \n', data.channels{:} ); fprintf('\n')
%}   

%% 
parser = inputParser();
varargin = ieParamFormat(varargin);

parser.addRequired('filename', @ischar);
parser.addParameter('datatype', 'radiance', @ischar);

parser.parse(filename, varargin{:});
filename = parser.Results.filename;
dataType = parser.Results.datatype;

%%

switch dataType
    case "radiance"
        output = piEXR2Mat(filename, 'Radiance');
    case "zdepth"
        output = piEXR2Mat(filename, 'Pz');
    case "depth"
        XDepthMap = piEXR2Mat(filename, 'Px');
        YDepthMap = piEXR2Mat(filename, 'Py');
        ZDepthMap = piEXR2Mat(filename, 'Pz');
        
        output = sqrt(XDepthMap.^2+YDepthMap.^2+ZDepthMap.^2);
    case "3dcoordinates"
        output(:,:,1) = piEXR2Mat(filename, 'Px');
        output(:,:,2) = piEXR2Mat(filename, 'Py');
        output(:,:,3) = piEXR2Mat(filename, 'Pz');

    case "material" % single channel
        output = piEXR2Mat(filename, 'MaterialId');
    case "normal"
        output(:,:,1) = piEXR2Mat(filename, 'Nx');
        output(:,:,2) = piEXR2Mat(filename, 'Ny');
        output(:,:,3) = piEXR2Mat(filename, 'Nz');
    case "albedo"
        % to add; only support rgb for now, spectral albdeo needs to add;
    case "instance" % single channel
        output = piEXR2Mat(filename, 'InstanceId');
    otherwise
        error('Datatype not supported. \n%s', 'Supported datatypes are: "radiance", "zdepth", "3dcoordinates", "material", "normal";')
end


end

File: src/pbrt/matlab/piReadEXR.cpp
#include"piReadEXR.h"
#include<cmath>
//读取深度时如果读不到就cout，跳过，不输入深度
#include"ieParamFormat.h"
#include "piEXR2Mat.h"
Energy piReadEXR(string filename, string datatype){
    // filename = ieParamFormat(filename); 这东西不能标准化，标准之后就找不到了
    datatype = ieParamFormat(datatype);
    //所有输出都可以先被统一成三通道，所以output就用energy了
    Energy output;
    if("radiance" == datatype){
        output = piEXR2Mat(filename,"Radiance");
    }else    if("zdepth" == datatype){
        output = piEXR2Mat(filename,"Pz");
    }else     if("depth" == datatype){
        Energy XDepthMap = piEXR2Mat(filename,"Px");
        Energy YDepthMap = piEXR2Mat(filename,"Py");
        Energy ZDepthMap = piEXR2Mat(filename,"Pz");
        //if no PZ ,break
        if(ZDepthMap.size() == 0){
            cout<<"no PZ"<<endl;
            return output;
        }
        vector<vector<double>> temp1;
        for(int i =0;i<XDepthMap.at(0).size();i++){
            vector<double> temp2;
            for(int j =0;j<XDepthMap.at(0).at(0).size();j++){
                double x = XDepthMap.at(0).at(i).at(j);
                double y = YDepthMap.at(0).at(i).at(j);
                double z = ZDepthMap.at(0).at(i).at(j);
                temp2.push_back(sqrt(x*x+y*y+z*z));
            }
            temp1.push_back(temp2);
        }
        output.push_back(temp1);
    }else    if("3dcoordinates" == datatype){
        // output = piReadEXR(filename,"Radiance");
        //这个有点难做，暂时不实现，因为用不到
    } else{
        cout<<"no invalid input in piReadEXR"<<endl;
    }
    return output;
    //    case "material" % single channel
    //     output = piEXR2Mat(filename, 'MaterialId');
    // case "normal"
    //     output(:,:,1) = piEXR2Mat(filename, 'Nx');
    //     output(:,:,2) = piEXR2Mat(filename, 'Ny');
    //     output(:,:,3) = piEXR2Mat(filename, 'Nz');
    // case "albedo"
    //     % to add; only support rgb for now, spectral albdeo needs to add;
    // case "instance" % single channel
    //     output = piEXR2Mat(filename, 'InstanceId');
}